hal/armv8m/exceptions.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Exception handling
+ *
+ * Copyright 2017, 2022 Phoenix Systems
+ * Author: Pawel Pisarczyk, Jakub Sejdak, Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "../exceptions.h"
+#include "../cpu.h"
+#include "../console.h"
+#include "../string.h"
+#include "config.h"
+
+
+void hal_exceptionsDumpContext(char *buff, exc_context_t *ctx, int n)
+{
+	static const char *mnemonics[] = {
+		"0 #InitialSP", "1 #Reset", "2 #NMI", "3 #HardFault",
+		"4 #MemMgtFault", "5 #BusFault", "6 #UsageFault", "7 #SecureFault",
+		"8 #", "9 #", "10 #", "11 #SVC",
+		"12 #Debug", "13 #", "14 #PendSV", "15 #SysTick"
+	};
+	size_t i = 0;
+
+	n &= 0xf;
+
+	hal_strcpy(buff, "\nException: ");
+	i = sizeof("\nException: ") - 1;
+	hal_strcpy(&buff[i], mnemonics[n]);
+	i += hal_strlen(mnemonics[n]);
+
+	i += hal_i2s("\n r0=", &buff[i], ctx->r0, 16, 1);
+	i += hal_i2s("  r1=", &buff[i], ctx->r1, 16, 1);
+	i += hal_i2s("  r2=", &buff[i], ctx->r2, 16, 1);
+	i += hal_i2s("  r3=", &buff[i], ctx->r3, 16, 1);
+
+	i += hal_i2s("\n r4=", &buff[i], ctx->r4, 16, 1);
+	i += hal_i2s("  r5=", &buff[i], ctx->r5, 16, 1);
+	i += hal_i2s("  r6=", &buff[i], ctx->r6, 16, 1);
+	i += hal_i2s("  r7=", &buff[i], ctx->r7, 16, 1);
+
+	i += hal_i2s("\n r8=", &buff[i], ctx->r8, 16, 1);
+	i += hal_i2s("  r9=", &buff[i], ctx->r9, 16, 1);
+	i += hal_i2s(" r10=", &buff[i], ctx->r10, 16, 1);
+	i += hal_i2s(" r11=", &buff[i], ctx->r11, 16, 1);
+
+	i += hal_i2s("\nr12=", &buff[i], ctx->r12, 16, 1);
+	i += hal_i2s("  sp=", &buff[i], (u32)ctx - 17 * 4, 16, 1);
+	i += hal_i2s("  lr=", &buff[i], ctx->lr, 16, 1);
+	i += hal_i2s("  pc=", &buff[i], ctx->pc, 16, 1);
+
+	i += hal_i2s("\npsp_s=", &buff[i], ctx->psp_s, 16, 1);
+	i += hal_i2s(" psr=", &buff[i], ctx->psr, 16, 1);
+	i += hal_i2s(" exr=", &buff[i], ctx->excret, 16, 1);
+	i += hal_i2s(" bfa=", &buff[i], *(u32 *)0xe000ed38, 16, 1);
```

Comment:
Please look at https://github.com/phoenix-rtos/phoenix-rtos-kernel/commit/c6f7d6c7355df92810e1116eaa93491954f94425 and https://github.com/phoenix-rtos/phoenix-rtos-kernel/commit/4c6ad6160c37bd1193c4357b3d43c160f8b7aad3